include/evmc/evmc.h

```diff
@@ -687,6 +687,71 @@ typedef struct evmc_result (*evmc_execute_fn)(struct evmc_instance* instance,
                                               size_t code_size);
 
 
+/** The opaque type representing a Client-side tracer object. */
+struct evmc_tracer_context;
+
+/**
+ * The callback to trace instructions execution in an EVM.
+ *
+ * This function informs the Client what instruction has been executed in the EVM implementation
+ * and what are the results of executing this particular instruction.
+ * The message level information (like call depth, destination address, etc.) are not provided here.
+ * This piece of information can be acquired by inspecting messages being sent to the EVM in
+ * ::evmc_execute_fn and the results of the messages execution.
+ *
+ * @param context                The pointer to the Client-side tracing context. This allows to
+ *                               implement the tracer in OOP manner.
+ * @param step                   The instruction counter: number of instructions executed.
+ *                               This counter starts from 0 for every message passed to
+ *                               ::evmc_execute_fn.
```

Comment:
Why call this `step` instead of the more conventional `programCounter`? Is this to facilitate ewasm?

Reply:
See https://github.com/ethereum/evmc/pull/32#discussion_r188351911

Reply:
This is not the program counter. Program counter points the instruction being executed, so it's `code_offset` in this case. The `step` is the trace entry index. 

I'm ok with renaming any of those if you have better ideas for names.

Reply:
How about `tracing_step` or `trace_index` ? Why is it needed?

Reply:
Precisely I wanted the "instruction counter" per current message (i.e. EVM call). This is not not needed (because the Host can count tracing callback calls on its own) *UNLESS* the VM will not report every instruction. 
I propose to rename it to `instruction_counter` or remove it for now.

Reply:
I think that's why `trace_counter`/`trace_step`/`trace_index` is the appropriate name.

I'm not sure this is a useful field though.

Reply:
I mean this case where VM only traces storage instructions:
```
PUSH 0
SLOAD
PUSH 9
SSTORE
```

When counting the number of traces you should get:
```
(0, SLOAD, ...)
(1, SSTORE, ...)
```

When counting number of instructions you should get:
```
(1, SLOAD, ...)
(3, SSTORE, ...)
```

In the second case the Host can figure out that some executed instructions are missing in the trace. In the first case the counter is redundant as the Host can count it itself.

I'm going to remove the param.

Reply:
> I mean this case where VM only traces storage instructions:

You mean a specific VM won't support tracing every instruction?

I have no problem keeping it, but `step` seems to be rather confusing. Consider some of the suggestions above.